src/tslintMulti.ts

```diff
@@ -0,0 +1,155 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as ts from "typescript";
+
+import {
+    DEFAULT_CONFIG,
+    findConfiguration,
+    findConfigurationPath,
+    getRelativePath,
+    getRulesDirectories,
+    loadConfigurationFromPath,
+} from "./configuration";
+import { EnableDisableRulesWalker } from "./enableDisableRules";
+import { findFormatter } from "./formatterLoader";
+import { IFormatter } from "./language/formatter/formatter";
+import { RuleFailure } from "./language/rule/rule";
+import { TypedRule } from "./language/rule/typedRule";
+import { getSourceFile } from "./language/utils";
+import { IMultiLinterOptions, LintResult } from "./lint";
+import { loadRules } from "./ruleLoader";
+import { arrayify } from "./utils";
+
+/**
+ * Linter that can lint multiple files in consecutive runs.
+ */
+class MultiLinter {
+
+    public static VERSION = "3.14.0";
+
+    public static findConfiguration = findConfiguration;
+    public static findConfigurationPath = findConfigurationPath;
+    public static getRulesDirectories = getRulesDirectories;
+    public static loadConfigurationFromPath = loadConfigurationFromPath;
+
+    private failures: RuleFailure[] = [];
+
+    /**
+     * Creates a TypeScript program object from a tsconfig.json file path and optional project directory.
+     */
+    public static createProgram(configFile: string, projectDirectory?: string): ts.Program {
+        if (projectDirectory === undefined) {
+            const lastSeparator = configFile.lastIndexOf("/");
+            if (lastSeparator < 0) {
+                projectDirectory = ".";
+            } else {
+                projectDirectory = configFile.substring(0, lastSeparator + 1);
+            }
+        }
+
+        const {config} = ts.readConfigFile(configFile, ts.sys.readFile);
+        const parsed = ts.parseJsonConfigFileContent(config, {readDirectory: ts.sys.readDirectory}, projectDirectory);
+        const host = ts.createCompilerHost(parsed.options, true);
+        const program = ts.createProgram(parsed.fileNames, parsed.options, host);
+
+        return program;
+    }
+
+    /**
+     * Returns a list of source file names from a TypeScript program. This includes all referenced
+     * files and excludes declaration (".d.ts") files.
+     */
+    public static getFileNames(program: ts.Program): string[] {
+        return program.getSourceFiles().map(s => s.fileName).filter(l => l.substr(-5) !== ".d.ts");
+    }
+
+    constructor(private options: IMultiLinterOptions, private program?: ts.Program) {
+        // Empty
+    }
+
+    public lint(fileName: string, source: string, configuration: any = DEFAULT_CONFIG): void {
```

Comment:
I'm not particularly keen on extending the scope of this PR, but let me say this:

> this function should return the formatted output just for the file currently being linted

This is *exactly* what the (Single)Linter did so far: Lint one file and return the (formatted) result.
Adopting this behavior would exactly contradict the intention of this PR: To separate linting from returning the formatted output (once for all files linted so far).

> if being used as part of a build, lint failures could start being displayed before all files are linted

I'm afraid this would require adopting my [second proposed solution](https://github.com/palantir/tslint/issues/656#issuecomment-238471972), while this PR covers the first proposed solution. Otherwise this would be incompatible with any formatter that requires header and footer.

If you're using console output, you could still achieve this by instantiating one `MultiLinter` per file (which is what the `(Single)Linter` would do). Returning *formatted* output for each file *and also* once at the end would be redundant (and a waste of computation time, if you ask me).

> The getResult() method can still exist as is, to provide a way to get all results at once

If the result wasn't formatted, I wouldn't mind, but as it is, this would be redundant work. In general, I agree that linting should be separated from the result which should be separated from the formatting, but this is out of the scope of this PR. If we change too many things at once, we will very likely introduce bugs.

> Another thing to consider here is that source is actually optional if a program was provided before (see #1096)

You're right, it needs to be optional (unfortunately the API is already quite confusing and this will add to it).